### PR TITLE
[REFACTOR] Refactoring Code - Services, Controllers and Routes createade and Utils and Webhooks updated

### PR DESCRIPTION
**Pull request by BigSamu**
Issue #32 is addressed on this PR. PR opened for review and merge with base code

### CODE REVIEW

File: src/controllers/changeset.controllers.js [hunk 1]
```diff
@@ -132,28 +132,25 @@ export const createOrUpdateChangesetFile = async (payload) => {
...
-        headOwner,
-        headRepo,
-        headBranch,
-        changesetFilePath,
-        commitMessage,
-        changesetFile?.sha
-      );
-    } else {
-      console.log("No changeset file to delete.");
-    }
```
Comment (BigSamu) [inline src/controllers/changeset.controllers.js]: @JohnathonBowers, why was this logic changed? Now I think, it would be better to move the logic of retrieving the file and check if exist inside the `deleteFileByPath()` function. In that case, the `sha` parameter will note required here.

**Review by BigSamu — COMMENTED** [review]
@JohnathonBowers, I sent you many reviews in your last commits. I would appreciate your check!

File: src/controllers/changeset.controllers.js (hunk 1 above)
Comment (JohnathonBowers) [reply src/controllers/changeset.controllers.js]: I changed the logic in order to avoid an unnecessary API call. As it was, we checked to see if the file existed, and, whether or not it existed, we still called the `deleteFile` API endpoint. By adding an `if` statement here, we don't need to rely on the `deleteFile` API call returning with a 404 error .

File: src/services/file.services.js [hunk 2]
```diff
@@ -167,7 +167,7 @@ export const deleteFileByPath = async (
     });
     console.log(`File '${path}' deleted successfully.`);
   } catch (error) {
-    if(error.status === 404) {
+    if(error.status === 404 || error.status === 422) {
```
Comment (BigSamu) [inline src/services/file.services.js:171]: Why the 422 code is here? A 422 code means `Unprocessable Content`, which is related to when you send a request that has data that cannot be processed accordingly (for example, you are waiting for a body with string values, and you send numbers). Here is a [link](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422) with details.

**Review by BigSamu — COMMENTED** [review]
Why the 422 code is here? A 422 code means `Unprocessable Content`, which is related to when you send a request that has data that cannot be processed accordingly (for example, you are waiting for a body with string values, and you send numbers). Here is a [link](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422) with details.

**Review by BigSamu — COMMENTED** [review]
Ok this. But bear in mind that the private key we are using in the .env will be possibly moved to a secure storage, such as AWS Secrets Manager. This is the way to go to secure `.pem` files. Take a look at this [link](https://medium.com/@munteanu210/what-is-a-pem-file-navigating-its-creation-use-and-conversion-in-ssl-a7ac689867ae) to understand what is a pem file.

File: src/main.js [hunk 3]
```diff
@@ -3,7 +3,6 @@
 // ---------------------------------------------------
 
 import express from "express";
-import fs from "fs";
```
Comment (BigSamu) [inline src/main.js:7]: Ok this. But bear in mind that the private key we are using in the .env will be possibly moved to a secure storage, such as AWS Secrets Manager. This is the way to go to secure `.pem` files. Take a look at this [link](https://medium.com/@munteanu210/what-is-a-pem-file-navigating-its-creation-use-and-conversion-in-ssl-a7ac689867ae) to understand what is a pem file.

File: src/utils/validators.utils.js [hunk 4]
```diff
@@ -5,25 +5,13 @@ import {
...
-export const isValidChangelogSection = (changelogSection) => {
-  if (!changelogSection) {
-    throw new InvalidChangelogHeadingError();
-  }
-  if (changelogSection[0].trim().length === 0) {
-    throw new EmptyChangelogSectionError();
-  }
-  return true;
-};
-
```
Comment (BigSamu) [inline src/utils/validators.utils.js]: @JohnathonBowers,

Here, we need to revert to what you did. The idea of the validators is to handle the errors and throw them in one file. That is why I put all import of errors in the `validators.utils.js` file. In the future, I want to apply a factory pattern to enhance the code with the errors, and for that reason I want to handle all of them in one location

**Review by BigSamu — COMMENTED** [review]
In function `extractChangelogEntries` we should not throw errors. That logic should be managed by the function `isValidChangelogSection`. Please revert this change and take a look this [review](https://github.com/BigSamu/OpenSearch-bot/commit/855bc0bf3aca1ff227d1f19b3a159c83922867f8#commitcomment-135977760)

**Review by BigSamu — COMMENTED** [review]
Please review this [commit](https://github.com/BigSamu/OpenSearch-bot/commit/0cac4ae0b438f096063b31fde817b720350c4b94#comments) 

In function extractChangelogEntries we should not throw errors. That logic should be managed by the function isValidChangelogSection. Please revert this change and take a look this [review](https://github.com/BigSamu/OpenSearch-bot/commit/855bc0bf3aca1ff227d1f19b3a159c83922867f8#commitcomment-135977760)

**Comment by BigSamu** [comment]
@JohnathonBowers,

Question regarding commit [0d3b10d](https://github.com/BigSamu/OpenSearch-bot/commit/0d3b10d85251d2b5a2889277fef7dce1c2ea09c2#comments). What was this merge about? I did the merges on Friday with the main branch. The question is, are there no conflicts currently with that branch?

File: src/utils/changelog.utils.js [hunk 5]
```diff
@@ -47,15 +48,19 @@ export const extractChangelogEntries = (
...
   processChangelogLine
 ) => {
-  // Match the changelog section using the defined regex
-  // Output -> Array of length 2:
-  // changelogSection[0]: Full regex match including '## Changelog' and following content.
-  // changelogSection[1]: Captured content after '## Changelog', excluding the heading itself.
-  // Throw error if '## Changelog' header is missing or malformed
-  const changelogSection = prDescription.match(CHANGELOG_SECTION_REGEX);
-
-  // Validate the changelog section and extract changelog entries
```
Comment (BigSamu) [inline src/utils/changelog.utils.js]: As I comment in other reviews, you should not send errors in the extractChangelogEntries() function. Instead a validator function should be imported to manage that.

File: src/services/auth.services.js [hunk 6]
```diff
@@ -15,13 +15,11 @@ import{
  * @throws {Error} - If an error occurs while obtaining the installation ID.
  */
 export const getOcktokitClient = async (owner, repo) => {
-
-  const ghApp = new App({
-    appId: GITHUB_APP_IDENTIFIER,
-    privateKey: GITHUB_APP_PRIVATE_KEY,
-  });
-
```
Comment (BigSamu) [inline src/services/auth.services.js]: Why did you move the ghApp instatntiation inside the try/catch block? At least for me, `try/catch` block should be easier to read. I would ask @ashwin-pc for better practices on this.

File: src/controllers/changeset.controllers.js [hunk 7]
```diff
@@ -140,6 +140,5 @@ export const createOrUpdateChangesetFile = async (payload) => {
     );
     // Clear skip label if exists
     await removeLabel(baseOctokit, baseOwner, baseRepo, prNumber, SKIP_LABEL);
-    throw new Error("Error creating or updating changeset file.");
```
Comment (BigSamu) [inline src/controllers/changeset.controllers.js]: But we need to throw an error here or not? How we show in the PR a message below saying that Merge cannot be processed given this error. Maybe throwing an error is not the solution, but we should show that the checks did not pass. For example, take a look below:

image

In this case we have an error but checks are saying that reviews are passing. We should send with an octokit that a check was not passed, I think. Can you take a look at that?

File: src/services/file.services.js (hunk 2 above)
Comment (JohnathonBowers) [reply src/services/file.services.js:171]: I added the 422 code here because I was getting this error when the program tried to delete a file that didn't exist. Since the `if` block only checked for a 404 error, the program continued executing when it should have terminated.

Screenshot 2023-12-31 at 12 02 14 PM

**Comment by JohnathonBowers** [comment]
In the [discussion](https://github.com/BigSamu/OpenSearch-bot/commit/4cf1d8d86428e050fad7626be3de3a7d770dcb66#comments) on commit [4cf1d8d](https://github.com/BigSamu/OpenSearch-bot/commit/4cf1d8d86428e050fad7626be3de3a7d770dcb66), @BigSamu added this comment:

> But we need to throw an error here or not? How we show in the PR a message below saying that Merge cannot be processed given this error. Maybe throwing an error is not the solution, but we should show that the checks did not pass. For example, take a look below:
> 
> image
> In this case we have an error but checks are saying that reviews are passing. We should send with an octokit that a check was not passed, I think. Can you take a look at that?

I responded:
> If we do throw an error here, we need to add additional error handling in the `github.webhooks.js` file. As it was, when I was trying to set up my local environment and figure out how to configure my settings to get everything to work properly, the thrown error here was making it difficult to identify the source of the error, especially since it was only printing a generic message to the console.

He responded:
> You have a point. But that error, in particular, was intended to throw a problem in the check. The middleware then catches the error I sent and throws a new error you were seeing. What we need is to show a check error. We can investigate how to properly show the error sent by the middleware in something more concise. Can you review this?

I'm moving this comment thread out into the body of the PR so it can be more visible.

**Comment by JohnathonBowers** [comment]
> @JohnathonBowers,
> 
> Question regarding commit [0d3b10d](https://github.com/BigSamu/OpenSearch-bot/commit/0d3b10d85251d2b5a2889277fef7dce1c2ea09c2#comments). What was this merge about? I did the merges on Friday with the main branch. The question is, are there no conflicts currently with that branch?

I had made some changes on my local tracking branch before you merged `main` into the upstream refactor branch. I needed to merge these upstream changes into my local tracking branch so that any code I pushed to the upstream branch wouldn't create conflicts. To answer your question, everything is in sync now.

File: src/main.js (hunk 3 above)
Comment (JohnathonBowers) [reply src/main.js:7]: I removed this import only because we weren't using it in the file. I'm fine with adding it back.

File: src/utils/validators.utils.js (hunk 4 above)
Comment (JohnathonBowers) [reply src/utils/validators.utils.js]: The validator function was not working as expected. The first check here: `if(!prDescription)` checks for a completely blank PR description. We did not have error handling for that scenario previously, and it was causing the app to behave unexpectedly. I can add comments on the other checks in this file, but, in brief, the validator function, at the point it was called, was returning false positives. The information it would need to see if a `## Changelog` section is empty is not available until after this code block:

```
// Process each line and filter out valid changelog entries
    const changelogEntries = changelogSection[0]
      .split("\n")
      .reduce((acc, line) => {
        const { entries, state } = acc;
        const processed = processChangelogLine(line, state);
        if (processed.line) entries.push(processed.line.trim());
        return { entries, state: processed.state };
      }, initialAcc).entries;
```

For reference, here is the validator function we were using:

```
export const isValidChangelogSection = (changelogSection) => {
  if (!changelogSection) {
    throw new InvalidChangelogHeadingError();
  }
  if (changelogSection[0].trim().length === 0) {
    throw new EmptyChangelogSectionError();
  }
  return true;
};
```

However, an empty changelog section at this point in the process looks like this:

```
  '## Changelog',
  '',
  index: 0,
  input: '## Changelog',
  groups: undefined
```

Checking if `changelogSection[0].trim().length === 0` at this point will return false, which will allow the program to continue executing rather than throwing an error, as we would want it to. My suggestion is that, if we want to keep the validator function, we should check the length of index 1 rather than index 0 in `changelogSection`.

File: src/main.js (hunk 3 above)
Comment (BigSamu) [reply src/main.js:7]: No need to add it back. I'm just letting you know that maybe we can use it later. You can resolve this conversation!

File: src/controllers/changeset.controllers.js (hunk 7 above)
Comment (JohnathonBowers) [reply src/controllers/changeset.controllers.js]: If we do throw an error here, we need to add additional error handling in the github.webhooks.js file. As it was, when I was trying to set up my local environment and figure out how to configure my settings to get everything to work properly, the thrown error here was making it difficult to identify the source of the error, especially since it was only printing a generic message to the console.

Comment (JohnathonBowers) [reply src/controllers/changeset.controllers.js]: From @BigSamu :

> You have a point. But that error, in particular, was intended to throw a problem in the check. The middleware then catches the error I sent and throws a new error you were seeing. What we need is to show a check error. We can investigate how to properly show the error sent by the middleware in something more concise. Can you review this?

Comment (BigSamu) [reply src/controllers/changeset.controllers.js]: You have a point. But that error, in particular, was intended to throw a problem in the check. The middleware then catches the error I sent and throws a new error you were seeing. What we need is to show a check error. We can investigate how to properly show the error sent by the middleware in something more concise. Can you review this?

File: src/services/auth.services.js (hunk 6 above)
Comment (JohnathonBowers) [reply src/services/auth.services.js]: I moved ghApp instantiation inside of the try/catch block so that errors instantiating the app would be caught and propagated up the call stack. When I was setting up the app in my local environment and trying to get my private key properly formatted, having the ghApp instantiation outside of the try/catch block made it difficult to debug.

Comment (BigSamu) [reply src/services/auth.services.js]: But why? Should it not make a big difference? What was the specific problem you had with that?

Comment (JohnathonBowers) [reply src/services/auth.services.js]: From @BigSamu :

> But why? Should it not make a big difference? What was the specific problem you had with that?

File: src/controllers/changeset.controllers.js (hunk 1 above)
Comment (BigSamu) [reply src/controllers/changeset.controllers.js]: Can you move the logic inside the deleteFileByPath() function?

File: src/services/file.services.js (hunk 2 above)
Comment (BigSamu) [reply src/services/file.services.js:171]: Here, it says we are passing a `sha` in the wrong format. Make sense. The original code is wrong I see. We always need to supply a correct sha to delete a file. We should not call the octokit delete function if the getFileContent function does not find a file. Can you correct this?

File: src/utils/validators.utils.js (hunk 4 above)
Comment (BigSamu) [reply src/utils/validators.utils.js]: Ok, so in this case we need to find a proper way to run the validator, or create two of them. But we need to allocate all errors inside the validator functions to keep consistency to have all these custom errors in one place.

**Comment by BigSamu** [comment]
> > @JohnathonBowers,
> > Question regarding commit [0d3b10d](https://github.com/BigSamu/OpenSearch-bot/commit/0d3b10d85251d2b5a2889277fef7dce1c2ea09c2#comments). What was this merge about? I did the merges on Friday with the main branch. The question is, are there no conflicts currently with that branch?
> 
> I had made some changes on my local tracking branch before you merged `main` into the upstream refactor branch. I needed to merge these upstream changes into my local tracking branch so that any code I pushed to the upstream branch wouldn't create conflicts. To answer your question, everything is in sync now.

Perfect!

File: src/controllers/changeset.controllers.js (hunk 1 above)
Comment (JohnathonBowers) [reply src/controllers/changeset.controllers.js]: Good idea. I just created an issue for this in our repository.

File: src/services/file.services.js (hunk 2 above)
Comment (JohnathonBowers) [reply src/services/file.services.js:171]: This scenario will be addressed in the refactor of the `deleteFileByPath()` function.

File: src/utils/validators.utils.js (hunk 4 above)
Comment (JohnathonBowers) [reply src/utils/validators.utils.js]: Sounds good. I've created an issue in the repo for adding a revised validator function(s).

File: src/services/auth.services.js (hunk 6 above)
Comment (JohnathonBowers) [reply src/services/auth.services.js]: I created an issue in our repo to add a separate try/catch block to handle any problems with instantiating `ghApp`.

File: src/utils/changelog.utils.js (hunk 5 above)
Comment (JohnathonBowers) [reply src/utils/changelog.utils.js]: I've opened an issue to create an updated validator function(s) that will handle errors.

File: src/controllers/changeset.controllers.js (hunk 7 above)
Comment (JohnathonBowers) [reply src/controllers/changeset.controllers.js]: I've opened an issue in our repo to address this.
